horovod/spark/common/util.py

```diff
@@ -171,24 +171,38 @@ def check_shape_compatibility(metadata, feature_columns, label_columns,
                     'model input at index {idx} with size {input}'
                     .format(col=col, feature=col_size, idx=idx, input=input_size))
 
-    if output_shapes is not None:
+    if label_shapes is not None:
         label_count = len(label_columns)
+        if label_count != len(label_shapes):
+            raise ValueError('Label column count {labels} must equal '
+                             'provided label shapes count {outputs}'
+                             .format(labels=label_count, outputs=len(label_shapes)))
+
+    label_count = len(label_columns)
+    if output_shapes is not None:
         if label_count != len(output_shapes):
             raise ValueError('Label column count {labels} must equal '
                              'model outputs count {outputs}'
                              .format(labels=label_count, outputs=len(output_shapes)))
 
-        for idx, col, output_shape in zip(range(label_count), label_columns, output_shapes):
-            col_size = metadata[col]['shape']
-            if col_size is None:
+    def _check_label_cols_size(target_shapes):
+        for _idx, _col, target_shape in zip(range(label_count), label_columns, target_shapes):
+            _col_size = metadata[_col]['shape']
+            if _col_size is None:
                 # When training directly on Parquet, we do not compute shape metadata
                 continue
 
-            output_size = abs(np.prod(output_shape))
-            if col_size != output_size:
+            target_size = abs(np.prod(target_shape))
+            if _col_size != target_size:
                 raise ValueError('Label column \'{col}\' with size {label} must equal that of the '
-                                 'model output at index {idx} with size {output}'
-                                 .format(col=col, label=col_size, idx=idx, output=output_size))
+                                 'model output and label shape at index {idx} with size {output}'
+                                 .format(col=_col, label=_col_size, idx=_idx, output=target_size))
+
+    if output_shapes is not None:
+        _check_label_cols_size(output_shapes)
```

Comment:
Seems this should only be done if `label_shapes is None`, right?  Otherwise, we will be forcing the model output size to equal the label size, which may not be correct (for these custom loss functions).